### PR TITLE
Implement equality semantics for async tagger tags.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/64797

This is necessary so that we can properly diff old/new tags and tell the editor the minimal change made.

### CODE REVIEW

File: src/EditorFeatures/Core.Wpf/InlineDiagnostics/InlineDiagnosticsTaggerProvider.cs [hunk 1]
```diff
@@ -121,5 +121,19 @@ protected internal override bool IncludeDiagnostic(DiagnosticData diagnostic)
...
+        /// way to reuse any of these currently.  Ideally we could do something similar to inline-hints where there's a
+        /// data tagger portion (which is async and has clean equality semantics), and then the UI portion which just
+        /// translates those data-tags to the UI tags.
+        /// <para>
+        /// Returning false here means we'll always end up regenerating all tags.  But hopefully there won't be that
+        /// many in a document to matter.
+        /// </para>
+        /// </summary>
+        protected override bool Equals(InlineDiagnosticsTag tag1, InlineDiagnosticsTag tag2)
+            => false;
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/InlineDiagnostics/InlineDiagnosticsTaggerProvider.cs:137]: @akhera99

File: src/EditorFeatures/Core.Wpf/LineSeparators/LineSeparatorTaggerProvider.cs [hunk 2]
```diff
@@ -114,5 +114,12 @@ private void OnFormatMappingChanged(object sender, FormatItemsEventArgs e)
                     context.AddTag(new TagSpan<LineSeparatorTag>(span.ToSnapshotSpan(snapshotSpan.Snapshot), tag));
             }
         }
+
+        /// <summary>
+        /// We create and cache a separator tag to use (unless the format mapping changes).  So we can just use identity
+        /// comparisons here.
+        /// </summary>
+        protected override bool Equals(LineSeparatorTag tag1, LineSeparatorTag tag2)
+            => tag1 == tag2;
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/LineSeparators/LineSeparatorTaggerProvider.cs:123]: many simple tags just need something like this.

File: src/EditorFeatures/Core.Wpf/StringIndentation/StringIndentationTag.cs [hunk 3]
```diff
@@ -31,5 +37,35 @@ internal class StringIndentationTag : BrushTag
...
+
+            return true;
+        }
+
+        private bool SpanEquals(SnapshotSpan snapshotSpan1, SnapshotSpan snapshotSpan2)
+        {
+            // Map span1 over to span2
+            var span1 = snapshotSpan1.TranslateTo(snapshotSpan2.Snapshot, _provider.SpanTrackingMode);
+            return span1.Span == snapshotSpan2.Span;
+        }
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/StringIndentation/StringIndentationTag.cs:69]: more complex tags have to normalize themselves to do the equality comparison properly.

File: src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource_ProduceTags.cs [hunk 4]
```diff
@@ -376,10 +376,9 @@ private IEnumerable<ITagSpan<TTag>> GetNonIntersectingTagSpans(IEnumerable<Snaps
             {
                 var snapshot = spansToInvalidate.First().Snapshot;
 
-                var tagSpansToInvalidate = new List<ITagSpan<TTag>>(
-                    spansToInvalidate.SelectMany(ss => oldTagTree.GetIntersectingSpans(ss)));
-
-                return oldTagTree.GetSpans(snapshot).Except(tagSpansToInvalidate, comparer: this);
+                return oldTagTree.GetSpans(snapshot).Except(
+                    spansToInvalidate.SelectMany(oldTagTree.GetIntersectingSpans),
+                    comparer: this);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource_ProduceTags.cs:381]: same as before, just no intermediary list computation.

File: src/Features/Core/Portable/InheritanceMargin/InheritanceMarginItem.cs [hunk 5]
```diff
@@ -58,6 +58,19 @@ namespace Microsoft.CodeAnalysis.InheritanceMargin
...
         }
 
+        public override int GetHashCode()
+            => throw ExceptionUtilities.Unreachable();
+
+        public override bool Equals(object? obj)
+            => obj is InheritanceMarginItem item && Equals(item);
+
+        public bool Equals(InheritanceMarginItem other)
+            => this.LineNumber == other.LineNumber &&
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/InheritanceMargin/InheritanceMarginItem.cs:68]: @Cosifne It would be good to not have LineNumber be part of hte data side of this.  It is positional, and shouldn't cause us to teardown/recreate tags.  In other words, if a member moves down one line, we should still be able to use its existing tags insteading of having to remove hte old one and create a new one.  

teh general pattern for this is to have ITagSpan in the tagger store the position, and the ITag itself it position-free.

Comment (jasonmalinowski) [reply src/Features/Core/Portable/InheritanceMargin/InheritanceMarginItem.cs:68]: Should we get a bug tracking this?

File: src/Features/Core/Portable/InheritanceMargin/InheritanceTargetItem.cs [hunk 6]
```diff
@@ -67,5 +67,19 @@ namespace Microsoft.CodeAnalysis.InheritanceMargin
...
             ProjectName = projectName;
         }
+
+        public override int GetHashCode()
+            => throw ExceptionUtilities.Unreachable();
+
+        public override bool Equals(object? obj)
+            => obj is InheritanceTargetItem item && Equals(item);
+
+        public bool Equals(InheritanceTargetItem other)
```
Comment (jasonmalinowski) [inline src/Features/Core/Portable/InheritanceMargin/InheritanceTargetItem.cs:77]: Can we make this a record?

File: src/VisualStudio/Core/Def/InheritanceMargin/InheritanceMarginTaggerProvider.cs [hunk 7]
```diff
@@ -150,5 +150,8 @@ protected override IEnumerable<SnapshotSpan> GetSpansToTag(ITextView? textView,
                     new InheritanceMarginTag(lineNumber, membersOnTheLineArray)));
             }
         }
+
+        protected override bool Equals(InheritanceMarginTag tag1, InheritanceMarginTag tag2)
+            => tag1.Equals(tag2);
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/InheritanceMargin/InheritanceMarginTaggerProvider.cs:155]: Why is this necessary? Wasn't the default still to call object.Equals which you already called?

File: src/EditorFeatures/Core/Structure/StructureTag.cs [hunk 8]
```diff
@@ -59,6 +61,25 @@ public StructureTag(AbstractStructureTaggerProvider tagProvider, BlockSpan block
...
+        {
+            return other != null &&
+                this.GuideLineHorizontalAnchorPoint == other.GuideLineHorizontalAnchorPoint &&
+                this.Type == other.Type &&
+                this.IsCollapsible == other.IsCollapsible &&
+                this.IsDefaultCollapsed == other.IsDefaultCollapsed &&
+                this.IsImplementation == other.IsImplementation &&
+                _tagProvider.SpanEquals(this.Snapshot, this.OutliningSpan, other.Snapshot, other.OutliningSpan) &&
+                _tagProvider.SpanEquals(this.Snapshot, this.HeaderSpan, other.Snapshot, other.HeaderSpan) &&
+                _tagProvider.SpanEquals(this.Snapshot, this.GuideLineSpan, other.Snapshot, other.GuideLineSpan);
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core/Structure/StructureTag.cs:80]: What worries mere here specifically for this tagger is this is assuming the editor is mapping spans the same way we are. If they have any custom mapping logic (or just pick a different tracking mode), we could be in a case where we get out of sync with the editor where our interval tree says one thing and the editor's tracking is something else.

File: src/EditorFeatures/Core/InlineHints/InlineHintDataTag.cs [hunk 9]
```diff
@@ -2,22 +2,70 @@
...
+
+        public override bool Equals(object? obj)
+            => obj is InlineHintDataTag tag && Equals(tag);
+
+        public bool Equals(InlineHintDataTag? other)
+        {
+            if (other is null)
+                return false;
+
+            // they have to match if they're going to change text.
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core/InlineHints/InlineHintDataTag.cs:48]: Was there a reason this was broken out versus a large && block like we usually do?

File: src/EditorFeatures/Core/InlineHints/InlineHintDataTag.cs [hunk 10]
```diff
@@ -2,22 +2,70 @@
...
+            if (this.Hint.ReplacementTextChange?.NewText != other.Hint.ReplacementTextChange?.NewText)
+                return false;
+
+            // Ensure both hints are talking about the same snapshot.
+            if (!_provider.SpanEquals(_snapshot, this.Hint.Span, other._snapshot, other.Hint.Span))
+                return false;
+
+            if (this.Hint.ReplacementTextChange != null &&
+                other.Hint.ReplacementTextChange != null &&
+                !_provider.SpanEquals(_snapshot, this.Hint.ReplacementTextChange.Value.Span, other._snapshot, other.Hint.ReplacementTextChange.Value.Span))
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core/InlineHints/InlineHintDataTag.cs:62]: Should this be grouped with the other replacement text change checks, or was this intentional to sort cheapest to most expensive?

File: src/EditorFeatures/Core/InlineHints/InlineHintDataTag.cs [hunk 11]
```diff
@@ -2,22 +2,70 @@
...
+        {
+            if (other is null)
+                return false;
+
+            // they have to match if they're going to change text.
+            if (this.Hint.ReplacementTextChange is null != other.Hint.ReplacementTextChange is null)
+                return false;
+
+            // the text change text has to match.
+            if (this.Hint.ReplacementTextChange?.NewText != other.Hint.ReplacementTextChange?.NewText)
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core/InlineHints/InlineHintDataTag.cs:53]: Given you're doing ? the earlier check is redundant.

File: src/EditorFeatures/Core/Diagnostics/DiagnosticsSquiggleTaggerProvider.cs [hunk 12]
```diff
@@ -126,5 +127,12 @@ protected internal override bool IncludeDiagnostic(DiagnosticData diagnostic)
                     return PredefinedErrorTypeNames.OtherError;
             }
         }
+
+        protected override bool Equals(IErrorTag tag1, IErrorTag tag2)
+        {
+            return tag1 is RoslynErrorTag errorTag1 &&
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core/Diagnostics/DiagnosticsSquiggleTaggerProvider.cs:133]: Should this do a hard cast, since we don't expect error tags to not be a RoslynErrorTag?

File: src/EditorFeatures/Core/Diagnostics/AbstractDiagnosticsAdornmentTaggerProvider.RoslynErrorTag.cs [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+            }
+
+            public override bool Equals(object? obj)
+                => Equals(obj as RoslynErrorTag);
+
+            public bool Equals(RoslynErrorTag? other)
+            {
+                return other != null &&
+                    this.ErrorType == other.ErrorType &&
+                    this._data.GetValidHelpLinkUri() == other._data.GetValidHelpLinkUri() &&
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core/Diagnostics/AbstractDiagnosticsAdornmentTaggerProvider.RoslynErrorTag.cs:60]: Is this expensive? I see there's some URI parsing but if the only input is looks to be the descriptor's HelpLinkUri can we just compare that?

File: src/EditorFeatures/Core.Wpf/InlineDiagnostics/InlineDiagnosticsTaggerProvider.cs [hunk 14]
```diff
@@ -121,5 +121,19 @@ protected internal override bool IncludeDiagnostic(DiagnosticData diagnostic)
                 return null;
             }
         }
+
+        /// <summary>
+        /// TODO: is there anything we can do better here? Inline diagnostic tags are not really data, but more UI
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core.Wpf/InlineDiagnostics/InlineDiagnosticsTaggerProvider.cs:126]: @akhera99?

**Review by jasonmalinowski — APPROVED** [review]
Approach seems generally fine but I do wish we'd just push the editor's tags for implement equality semantics than us doing it ourselves. Seems like it'd be better for other teams (C++, etc.) and also means if the editor ever adds additional properties we don't have to update, etc.

File: src/EditorFeatures/Core.Wpf/InlineDiagnostics/InlineDiagnosticsTaggerProvider.cs (hunk 14 above)
Comment (akhera99) [reply src/EditorFeatures/Core.Wpf/InlineDiagnostics/InlineDiagnosticsTaggerProvider.cs:126]: I think what the comment says makes sense, separating the data tagging and creating UI tags from the data tags. I would be happy to change that in a follow-up PR.

File: src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource_IEqualityComparer.cs [hunk 15]
```diff
@@ -2,23 +2,27 @@
...
 namespace Microsoft.CodeAnalysis.Editor.Tagging
 {
     internal abstract partial class AbstractAsynchronousTaggerProvider<TTag>
     {
         private partial class TagSource : IEqualityComparer<ITagSpan<TTag>>
         {
-            public bool Equals(ITagSpan<TTag> x, ITagSpan<TTag> y)
-                => x.Span == y.Span && EqualityComparer<TTag>.Default.Equals(x.Tag, y.Tag);
+            public bool Equals(ITagSpan<TTag>? x, ITagSpan<TTag>? y)
+                => x != null && y != null && x.Span == y.Span && _dataSource.Equals(x.Tag, y.Tag);
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource_IEqualityComparer.cs:16]: i found that too easy to miss.  this way (being abstract) forced all impls to have to consider what to do here.  :-/

File: src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.cs [hunk 16]
```diff
@@ -81,6 +82,16 @@ internal abstract partial class AbstractAsynchronousTaggerProvider<TTag> where T
         /// </summary>
         protected virtual TaggerDelay AddedTagNotificationDelay => TaggerDelay.NearImmediate;
 
+        /// <summary>
+        /// Comparer used to check if two tags are the same.  Used so that when new tags are produced, they can be
+        /// appropriately 'diffed' to determine what changes to actually report in <see cref="ITagger{T}.TagsChanged"/>.
+        /// </summary>
+        protected abstract bool Equals(TTag tag1, TTag tag2);
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.cs:89]: makes sense!

File: src/VisualStudio/Core/Def/InheritanceMargin/InheritanceMarginTaggerProvider.cs (hunk 7 above)
Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/InheritanceMargin/InheritanceMarginTaggerProvider.cs:155]: no.  there is no default.  i wanted an override in all cases to force consideration of the correct semantics :)

File: src/EditorFeatures/Core/Structure/StructureTag.cs (hunk 8 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Structure/StructureTag.cs:80]: so i'm not worried about that because the tagProvider *determiens* the tracking mode.  that's why i'm deferring to it to do the equality check.

File: src/EditorFeatures/Core/InlineHints/InlineHintDataTag.cs (hunk 10 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/InlineHints/InlineHintDataTag.cs:62]: intentional for that reason.

File: src/EditorFeatures/Core/Diagnostics/DiagnosticsSquiggleTaggerProvider.cs (hunk 12 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Diagnostics/DiagnosticsSquiggleTaggerProvider.cs:133]: yes.  that seems reasonable.

File: src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.cs (hunk 16 above)
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.cs]: done!

File: src/Features/Core/Portable/Common/TaggedText.cs [hunk 17]
```diff
@@ -83,6 +84,20 @@ internal TaggedText(string tag, string text, TaggedTextStyle style, string navig
 
         public override string ToString()
             => Text;
+
+        public bool Equals(TaggedText other)
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/Common/TaggedText.cs]: done!

File: src/Features/Core/Portable/InheritanceMargin/InheritanceTargetItem.cs (hunk 6 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/InheritanceMargin/InheritanceTargetItem.cs]: done!

File: src/EditorFeatures/Core/InlineHints/InlineHintDataTag.cs (hunk 11 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/InlineHints/InlineHintDataTag.cs:53]: the earlier null check is making sure they match in nulability.  i.e. that both are either null or not null.  it doesn't actually verify that they're not null :)

File: src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.cs [hunk 18]
```diff
@@ -58,7 +59,7 @@ internal abstract partial class AbstractAsynchronousTaggerProvider<TTag> where T
         /// created for a previous version of a document that are mapped forward by the async
         /// tagging architecture.  This value cannot be <see cref="SpanTrackingMode.Custom"/>.
         /// </summary>
-        protected virtual SpanTrackingMode SpanTrackingMode => SpanTrackingMode.EdgeExclusive;
+        public virtual SpanTrackingMode SpanTrackingMode => SpanTrackingMode.EdgeExclusive;
```
Comment (genlu) [inline src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.cs:62]: This is a breaking change for TS
http://index/?query=Microsoft.CodeAnalysis.Editor.TypeScript.Features.Classifier.SemanticClassificationTaggerProvider&rightProject=Microsoft.CodeAnalysis.TypeScript.EditorFeatures&file=Features%5CClassifier%5CSemanticClassificationTaggerProvider.cs&line=43

File: src/EditorFeatures/Core/Structure/StructureTag.cs [hunk 19]
```diff
@@ -59,6 +61,25 @@ public StructureTag(AbstractStructureTaggerProvider tagProvider, BlockSpan block
         public bool IsDefaultCollapsed { get; }
         public bool IsImplementation { get; }
 
+        public override int GetHashCode()
+            => throw ExceptionUtilities.Unreachable();
```
Comment (genlu) [inline src/EditorFeatures/Core/Structure/StructureTag.cs:65]: This is called in editor

https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_git/VS-Platform?path=/src/Editor/Text/Impl/Structure/StructureSpanningTree/StructureSpanningTree.cs&version=GBmain&line=308&lineEnd=309&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents

File: src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.cs (hunk 18 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.cs:62]: fixed in followup.

File: src/EditorFeatures/Core/Structure/StructureTag.cs (hunk 19 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Structure/StructureTag.cs:65]: fixed in followup.
